File: apps/api/app/core/vocabulary.py
"""What each cloud calls the things a customer reads about.

The columns are neutral and named after Azure's words, deliberately and for a
reason that has not changed: renaming them would make every stored capture
unreplayable (``DECISIONS.md`` §70). What a *customer* reads is a different
question, and the answer to that one is not negotiable — a CSPM telling an AWS
customer "this subscription is no longer connected" is a product that has not
noticed which cloud it is looking at.

So the identifiers stay Azure-flavoured and the sentences do not. This module is
the one place that knows the difference, and it holds nouns rather than
sentences: a sentence per provider would be two sentences to keep in step, and
the half nobody was looking at would drift.

Deliberately not on ``ProviderOnboarding``. These words are needed by the
scanner, the findings routes and the scan model — none of which is onboarding,
and all of which would otherwise import a connector to ask what a word is.
"""

from dataclasses import dataclass

from app.core.enums import Provider


@dataclass(frozen=True)
class Words:
    """One cloud's nouns, in the forms a sentence needs them."""

    #: The unit a scan reads: a subscription, an account, later a project.
    account: str
    accounts: str
    #: The trust boundary above it.
    boundary: str
    #: What the boundary's own reading is *of* — the directory on Azure, the
    #: organization on AWS. Distinct from ``boundary`` because a sentence says
    #: "read the tenant directory" and "read the organization", not "read the
    #: tenant organization".
    directory: str
    #: What the customer deploys to grant access, for the sentences that name it.
    artifact: str
    #: Where they deploy it.
    console: str


AZURE = Words(
    account="subscription",
    accounts="subscriptions",
    boundary="tenant",
    directory="tenant directory",
    artifact="scanner role",
    console="Azure Portal",
)

AWS = Words(
    account="account",
    accounts="accounts",
    boundary="organization",
    directory="organization",
    artifact="scanner stack",
    console="the AWS console",
)

# A provider with no entry gets Azure's words rather than an exception. This is
# a sentence, not a security decision: a missing noun should read slightly wrong
# rather than break the page a customer is trying to read.
_WORDS: dict[Provider, Words] = {Provider.AZURE: AZURE, Provider.AWS: AWS}


def words(provider: Provider | str | None) -> Words:
    """The nouns for one cloud."""
    if provider is None:
        return AZURE
    try:
        return _WORDS.get(Provider(provider), AZURE)
    except ValueError:
        return AZURE

File: apps/web/src/lib/vocabulary.ts
import type { Provider } from "@/lib/types";

/**
 * What each cloud calls the things a customer reads about.
 *
 * The API's field names keep Azure's vocabulary — `subscriptions`,
 * `subscription_id` — because the columns behind them do, and renaming those
 * would make every stored capture unreplayable (`DECISIONS.md` §70). What a
 * customer *reads* is a different question, and telling somebody with an AWS
 * connection that CloudGuard found "3 subscriptions" is a product that has not
 * noticed which cloud it is looking at.
 *
 * So identifiers stay Azure-flavoured and sentences do not. Nouns rather than
 * sentences, for the reason `setupCopy` holds overrides rather than a second
 * copy of the whole block: two sets of sentences drift, and the half that
 * drifts is the half nobody is looking at.
 */
export interface Words {
  /** The unit a scan reads. */
  account: string;
  accounts: string;
  /** Capitalised, for a column heading or the start of a sentence. */
  Account: string;
  Accounts: string;
  /** The trust boundary above it. */
  boundary: string;
  /** What the boundary's own reading is *of*. */
  directory: string;
  /** What the customer deploys, and where. */
  artifact: string;
  console: string;
}

const AZURE: Words = {
  account: "subscription",
  accounts: "subscriptions",
  Account: "Subscription",
  Accounts: "Subscriptions",
  boundary: "tenant",
  directory: "tenant directory",
  artifact: "scanner role",
  console: "Azure Portal",
};

const AWS: Words = {
  account: "account",
  accounts: "accounts",
  Account: "Account",
  Accounts: "Accounts",
  boundary: "organization",
  directory: "organization",
  artifact: "scanner stack",
  console: "the AWS console",
};

/**
 * The nouns for one cloud.
 *
 * An unknown provider gets Azure's words rather than throwing. This is a
 * sentence, not a security decision: a missing noun should read slightly wrong
 * rather than break the page somebody is trying to read.
 */
export function words(provider: Provider | string | undefined | null): Words {
  return provider === "aws" ? AWS : AZURE;
}
